### DISCUSSION TITLE
Substituting missing types

### DISCUSSION THREAD

**Discussion by JensNordenbro** [dotnet/runtime#41020] [General]
Hi!

Is it possible for a .NET Core developer to manually add types that are missing from in .net5 but present in .net framework in order to cope with compatility issues when using .netfw depenencies and allow for a gradual modernization of a code base? 

I do not know how Microsoft.Windows.Compatibility does it but I like how it separated what Microsoft wants to be the future API vs legacy API.

**Comment by danmoseley** [comment]
@JensNordenbro you would not be able to do that realistically without modifying the calling code, because the type needs to be in the same assembly identity -- eg., System.Core, and we already deploy an assembly with that identity. What you would have to do is modify the calling code to use your private implementation.

What types are you missing?

**Reply by JensNordenbro** [reply]
Right now it is 
**1. System.Windows.Forms.ContextMenu** 
2. System.Windows.Forms.DataGrid et al 

The first one I cannot substitue without replacing a deeply embedded (hard to replace) third party component and I dont think I will be able to convince the partner to port to .net 5 in due time. 

The second one I have more hopes to cope with since that code belongs to us.
 

If it was possible to add "substitues" for missing types in a way like below I could take care of it myself :) 
metacode:
```
Assembly a = typeof(System.Windows.Forms.Controls.DataGridView).Assembly;
a.AddNewType("System.Windows.Forms.ContextMenu", new MyAssembleNamespace.BinaryCompatibleReplacementToContextMenuControl());

```

**Reply by JensNordenbro** [reply]
related is https://github.com/dotnet/runtime/discussions/41088
